### DISCUSSION TITLE
Question about filament vs spool

### DISCUSSION THREAD

**Discussion by spyder007** [Donkie/Spoolman#338] [General]
As I just started this, it occurs to me that some of the fields in the `filament` seem to be more "spool-specific."  This runs into some very unique issues and I'm not sure how to get past it. I can see `spool weight` being a vendor property, and perhaps allow the filament to "override" the empty spool weight (maybe the vendor uses different spools for PLA vs ABS, etc).  #307 is, I believe, a feature request for exactly that.  

Likewise, perhaps `weight` can be set as a default on the filament, but, at the end of the day, every spool is a different weight.  Issue #299  highlights this.

Would anyone be opposed to something like this:

1.  Add `Empty Spool Weight` field to vendor (not required) and allow that value to "fall through" the filament to the spool.  If you don't set an empty weight at the filament level, we will use the empty weight at the vendor level.   This should solve #307 
2. Add `Initial Weight` field to the spool object.  It will initially inherit from the `Weight` field of the Filament, but would be able to be changed, to support over-laden rolls.  This fixes #299 and is also useful if you are transferring filament from bulk rolls.
3.  Change the spool math as follows:
       `Used Weight` - If used weight is entered=
          1. `remaining weight = initial weight - used weight`
          2. `measured weight = (initial weight - used weight) + empty spool weight`
     `Remaining Weight` - If remaining weight is entered
          1. `used weight = initial weight - remaining weight`
          2. `measured weight = remaining weight + empty spool weight`
     `Measured Weight` - If measured weight is entered
          1. `used weight = initial weight - (measured weight - empty spool weight)`
          2. `remaining weight = measured weight - empty spool weight`

I also generally wonder if any of this belongs at the filament level, or if it should go straight to the spool level.  Filament should just be properties of the filament, not properties of the spool.  If the above was implemented, the next steps could be:
1. Remove `Weight` and `Spool Weight` from the filament
2. Add `Empty Spool Weight` to the spool itself as an editable field that defaults to the empty spool weight from the vendor.  This would allow you to set spool weight based on the spool itself, not some value on the filament.

I realize that this means a little more data entry when entering a spool, particularly if your vendor uses a variety of spools.  But it separates the filament properties from the spool properties and generally makes the math easier.

**Comment by BigTrikerTom** [comment]
That would be a good approach.

**Comment by jeffalanmcclain** [comment]
My understanding is that Price and Weight is already a "default" for the Filament, and Spool over-rides.  This is certainly how the "hover over" help text for fields shows, however, price (when left blank on a spool) certainly doesn't seem to grab the filament value, and this is likely a form entry issue (blank / null vs true ZERO value).  But I would appreciate that being fixed as a first order priority and I think most of your points will be already addressd.

**Comment by amacleod** [comment]
Would it be reasonable to allow for negative values of "used weight"? The weight number is just an estimate, after all. If your scale is good, and the spool weight is well known, it can be a very good estimate, but an estimate nonetheless. For the most part, it seems like vendors err on the side of generosity: most of the "1kg spools" I have tracked carefully turned out to be in the range of 1005-1025g net weight. It sure would be convenient to say that a class of spools are nominally 1000g, but allow the remaining weight to be 1009g. Fudging the initial weight to 1009g would work, if the system allows it, but maybe allowing the used weight to be -9g could be simpler?

(edit to add: looks like you already considered the "fudge `initial_weight`" thing in https://github.com/Donkie/Spoolman/pull/339#discussion_r1566488384 -- if I am commenting on a done deal, please disregard 🙂)
